### PR TITLE
Clear cache when adding notes w/o sort

### PR DESCRIPTION
**Pull request by mscuthbert**
This was failing:

```
>>> c = chord.Chord('C4 E4 G4')
>>> c.isConsonant()
True

>>> c.add('C#4', runSort=False)
>>> c.isConsonant()
True
```

The cache was only being cleared during `runSort=True`, which is the default.  But needs to happen either side.

Also caches the sort status, so repeated Chord.sort(inPlace=True) should be much faster.

### CODE REVIEW

**Comment by mscuthbert** [comment]
Requesting review from @jacobtylerwalls -- I think this is an improvement, but it makes one thing cached (chord's sort status) that requires better informing from pitches and notes when one of them changes.  It might need some more cases.

File: music21/chord/__init__.py [hunk 1]
```diff
@@ -3989,6 +4025,7 @@ def semiClosedPosition(self, *, forceOctave=None, inPlace=False, leaveRedundantP
                     newRemainingPitches.append(p)
             remainingPitches = newRemainingPitches
 
+        c2.clearCache()
```
Comment (jacobtylerwalls) [inline music21/chord/__init__.py:4028]: The octave setter informs client, so I think you're okay--but belt and suspenders is fine.

File: music21/pitch.py [hunk 2]
```diff
@@ -2863,6 +2872,7 @@ def step(self, usrStr: StepName) -> None:
             self.spellingIsInferred = False
         else:
             raise PitchException(f'Cannot make a step out of {usrStr!r}')
+        self.informClient()
```
Comment (jacobtylerwalls) [inline music21/pitch.py:2875]: great!

**Review by jacobtylerwalls — APPROVED** [review]
Great, this is an improvement. I see why you wanted extra eyes.

Existing issue -- only one client is possible, so you can have situations like this. Not for this PR, but just mentioning:
```python
>>> n = note.Note('F')
>>> c1 = chord.Chord([n])
>>> c2 = chord.Chord([n])
>>> c1.sortAscending(inPlace=True)
>>> c2.sortAscending(inPlace=True)
>>> n.transpose(6, inPlace=True)
>>> c2._cache
{}
>>> c1._cache
{'isSortedAscendingDiatonic': True}
```

**Comment by mscuthbert** [comment]
you're right -- maybe better to keep `._client` private and just disable pylint any time we need to look at `other._client`.

**Comment by mscuthbert** [comment]
> ```python
> >>> n = note.Note('F')
> >>> c1 = chord.Chord([n])
> >>> c2 = chord.Chord([n])
> >>> c1.sortAscending(inPlace=True)
> >>> c2.sortAscending(inPlace=True)
> >>> n.transpose(6, inPlace=True)
> >>> c2._cache
> {}
> >>> c1._cache
> {'isSortedAscendingDiatonic': True}
> ```

oh... that is a thing.  Need to consider a bit more.

The original plan when Chris Ariza and I designed the sites system was that things like ._chordAttached would be a site for an object -- that sites wouldn't just be Streams that they were in but other contexts, including things like interpretive elements (tuning systems, etc.) --so that there were sites and "locations" (Streams) which are the only things (besides `None`) that we're storing in Sites right now.  Maybe that is the better option, since we have a lot of mechanisms for informing sites.

The same thing about having only one client for an object goes with Pitch objects as well.

File: music21/pitch.py [hunk 3]
```diff
@@ -897,7 +897,7 @@ def __init__(self, specifier: t.Union[int, str, float] = 'natural'):
         self.displayLocation = 'normal'
 
         # store a reference to the object that has this duration object as a property
-        self._client: t.Optional['Pitch'] = None
+        self.client: t.Optional['music21.note.Note'] = None
```
[the remark this answers is not in the document]
Comment (mscuthbert) [reply music21/pitch.py:900]: `_client` will remain private.

[the remark this answers is not in the document]
Comment (mscuthbert) [reply music21/pitch.py:900]: And ACK -- that was right before.

**Review by mscuthbert — COMMENTED** [review]
fixing the mess made last time.

**Pull request merged**